cli/azd/pkg/spin/spin.go

```diff
@@ -0,0 +1,104 @@
+package spin
+
+import (
+	"fmt"
+	"io"
+	"sync"
+	"time"
+
+	"github.com/mattn/go-colorable"
+	"github.com/theckman/yacspin"
+)
+
+// Default writer to std.out, with possibility to mock for tests
+var writer io.Writer = colorable.NewColorableStdout()
+
+// Spinner is a type representing an animated CLi terminal spinner.
+type Spinner struct {
+	spinner  *yacspin.Spinner
+	logMutex sync.Mutex
+}
+
+// Sets the title of the spinner.
+func (s *Spinner) Title(title string) {
+	if title != "" {
+		s.spinner.Message(title)
+	}
+}
+
+// Logs a message to standard output and pushes the spinner onto a new line.
+// Example:
+// Console output before, with spinner title set to "Doing things...":
+// > Doing things... X
+//
+// Console output after LogMessage("Step 1 completed."):
+// > Step 1 completed.
+// > Doing things... X
+func (s *Spinner) Println(message string) error {
+	if message != "" {
+		defer s.logMutex.Unlock()
+
+		s.logMutex.Lock()
+		err := s.spinner.Stop()
+		if err != nil {
+			return err
+		}
+
+		fmt.Fprintln(writer, message)
+
+		err = s.spinner.Start()
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Run renders the spinner while runFn is executing,
+// returning the error from executing runFn.
+// The spinner message is erased when the spinner is stopped.
+func (s *Spinner) Run(runFn func() error) error {
+	err := s.spinner.Start()
+	if err != nil {
+		return fmt.Errorf("starting spinner: %w", err)
+	}
+
+	// Can only error if the spinner is already stopped.
+	// nolint:errcheck
+	defer s.spinner.Stop()
+
+	return runFn()
+}
+
+// Starts the spinner. Only possible error is if the spinner is already running.
+func (s *Spinner) Start() error {
+	return s.spinner.Start()
```

Comment:
Hmm I can't actually think of a use case of where we would actually care about the `err`. We're asserting the state of the spinner by calling `Start()`. Does it matter that it was already started? 

The only contrived situation this could be of use is:
1. The spinner is a shared resource across multiple threads
2. It is somehow crucial that `Start()` and `Stop()` are called in a complete determistic order. This would currently give us a signal that something isn't working quite as expected.

which sounds like there should be other synchronization primitives to enforce ordering.

Reply:
Going to move forward with removing the `error` returned since the assertion vs. the action is a meaningless difference for us.